components/core/src/reducer/Operator.hpp

```diff
@@ -0,0 +1,69 @@
+#ifndef REDUCER_OPERATOR_HPP
+#define REDUCER_OPERATOR_HPP
+
+#include <memory>
+#include <set>
+#include <utility>
+
+#include "RecordGroup.hpp"
+#include "RecordGroupIterator.hpp"
+
+namespace reducer {
+enum class OperatorType : uint8_t {
+    Map,
+    GroupBy,
+    Reduce
+};
+
+enum class OperatorResultCardinality : uint8_t {
+    One,
+    Subset,
+    Input
+};
+
+/**
+ * Class implementing a generic Operator that operates on RecordGroup objects.
+ *
+ * Operators implement methods to accept inter-stage RecordGroups and intra-stage RecordGroups.
+ * In the former case, the operator acts like a combiner, combining records from a previous pipeline
+ * stage. In the latter case, the operator acts like a reducer, reducing pre-aggregated results from
+ * within a pipeline stage.
+ *
+ * Operators can be of type GroupBy, Map, or Reduce, though currently no Map type Operators are
+ * implemented.
+ */
+class Operator {
+public:
+    virtual ~Operator() = default;
+
+    virtual void
+    push_intra_stage_record_group(GroupTags const& tags, ConstRecordIterator& record_it)
```

Comment:
I could probably figure this out, but in the interest of time, why don't these methods take a RecordGroup directly?

Reply:
Since the iterator is a virtual class the only way to avoid dynamically allocating them is to have a RecordGroup have the iterator as a member and pass it out by reference. This means that if we pass in a RecordGroup we can't pass it in by const& because the iterator needs to be mutable. That wouldn't be a real problem because all of the RecordGroup implementations are really immutable views into some data, but it does look a bit weird.